src/terrain/generation/land_generator.cpp

```diff
@@ -195,16 +108,113 @@ LandGenerator::get_volume(int center[2][2], int Sxy, int Sz, int Dxy, int Dz) co
     }
 
     int z_max = rand() % (Dz + 1) + Sz - Dz / 2;
-    return {x_min, y_min, 0, x_max, y_max, z_max};
+    return {
+        x_min,
+        y_min,
+        0,
+        x_max,
+        y_max,
+        z_max,
+        stamp_material_id_,
+        stamp_color_id_,
+        elements_can_stamp_
+    };
 }
 
-void
-LandGenerator::next() {
-    current_sub_region++;
-    if (current_sub_region == get_num_stamps(data_[current_region])) {
-        current_region++;
-        current_sub_region = 0;
+std::set<std::pair<MaterialId, ColorId>>
+JsonToTile::read_elements_can_stamp(const Json::Value& data) {
+    std::set<std::pair<MaterialId, ColorId>> out;
+
+    for (const Json::Value& material_data : data["Can_Stamp"]) {
+        int E = material_data["E"].asInt();
+        if (material_data["C"].isInt()) {
+            int C = material_data["C"].asInt();
+            out.insert(std::make_pair(E, C));
+        } else if (material_data["C"].asBool()) {
+            out.insert(std::make_pair(E, -1));
+        }
+    }
+    return out;
+}
+
+FromRadius::FromRadius(const Json::Value& data) : JsonToTile(data) {
+    number_ = data["Radius"]["number"].asInt();
+    radius_ = data["Radius"]["radius"].asInt();
+
+    center_variance_ = data["DC"].asInt();
+}
+
+TileStamp
+FromRadius::operator()(int current_sub_region) const {
+    double distance = static_cast<double>(8 * radius_) / number_ * current_sub_region;
+    int side = static_cast<int>(distance) / 2 / radius_;
+    int x_center, y_center;
+
+    if (side == 0) {
+        x_center = -radius_;
+        y_center = -radius_ + std::fmod(distance, 2 * radius_);
+    } else if (side == 1) {
+        x_center = -radius_ + std::fmod(distance, 2 * radius_);
+        y_center = radius_;
+    } else if (side == 2) {
+        x_center = radius_;
+        y_center = radius_ - std::fmod(distance, 2 * radius_);
+    } else if (side == 3) {
+        x_center = radius_ - std::fmod(distance, 2 * radius_);
+        y_center = -radius_;
+    } else {
+        throw std::invalid_argument("Something went horribly wrong");
```

Comment:
Don't use an exception just `LOG_CRITICAL` with an informative log msg and `abort()`.